datafusion-cli/src/object_storage.rs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{env, str::FromStr, sync::Arc};
+
+use datafusion::datasource::object_store::ObjectStoreProvider;
+use object_store::{aws::AmazonS3Builder, gcp::GoogleCloudStorageBuilder};
+use url::Url;
+
+#[derive(Debug, PartialEq, Eq, clap::ArgEnum, Clone)]
+pub enum ObjectStoreRegistrationOptions {
+    S3,
+    GCS,
+    Unsupported,
+}
+
+impl FromStr for ObjectStoreRegistrationOptions {
+    type Err = String;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        // clap::ArgEnum::from_str(input, true)
+        match input {
+            "s3" => Ok(ObjectStoreRegistrationOptions::S3),
+            "gcs" => Ok(ObjectStoreRegistrationOptions::GCS),
+            _ => Ok(ObjectStoreRegistrationOptions::Unsupported),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct DatafusionCliObjectStoreProvider {}
+
+/// ObjectStoreProvider for S3 and GCS
+impl ObjectStoreProvider for DatafusionCliObjectStoreProvider {
+    fn get_by_url(&self, url: &Url) -> Option<Arc<dyn object_store::ObjectStore>> {
+        if let (Ok(object_store_option), Some(host)) = (
+            ObjectStoreRegistrationOptions::from_str(url.scheme()),
+            url.host_str(),
+        ) {
+            match object_store_option {
+                ObjectStoreRegistrationOptions::S3 => {
+                    match AmazonS3Builder::from_env().with_bucket_name(host).build() {
+                        Ok(s3) => Some(Arc::new(s3)),
+                        Err(_) => None,
+                    }
```

Comment:
ideally I would like to write this as: 
```
AmazonS3Builder::from_env()
    .with_bucket_name(host).build()
    .map(Arc::new)
    .ok();
```
but getting the following type error: 
```
mismatched types
expected enum `Option<Arc<(dyn object_store::ObjectStore + 'static)>>`
found enum `Option<Arc<AmazonS3>>`
```